hazelcast/include/hazelcast/client/sql/sql_page.h

```diff
@@ -0,0 +1,177 @@
+         * @param column_name column name
+         * @return value of the column
+         *
+         * @throws hazelcast::client::exception::illegal_argument if a column
+         * with the given name is not found
+         * @throws boost::any_cast_exception if the type of the column type
+         * isn't assignable to the type \codeT\endcode
+         *
+         * @see sql_column_metadata#name()
+         * @see sql_column_metadata#type()
+         */
+        template<typename T>
+        boost::optional<T> get_object(const std::string& column_name) const
+        {
+            auto column_index = resolve_index(column_name);
+            return page_->get_column_value<T>(column_index, row_index_);
+        }
+
+        /**
+         * Gets the row metadata.
+         * @returns vector of column metadata, and returns empty vector if the
+         * result contains only an update count.
+         *
+         */
+        const sql_row_metadata& row_metadata() const;
+
+    private:
+        std::size_t row_index_;
+        const sql_page* page_;
+        const sql_row_metadata* row_metadata_;
+
+        std::size_t resolve_index(const std::string& column_name) const;
+    };
+
+    sql_page(std::vector<sql_column_type> column_types,
+             std::vector<column> columns,
+             bool last);
+
+    sql_page(sql_page&& rhs) noexcept;
+
+    sql_page(const sql_page& rhs) noexcept;
+
+    sql_page& operator=(sql_page&& rhs) noexcept;
+
+    sql_page& operator=(const sql_page& rhs) noexcept;
+
+    const std::vector<sql_column_type>& column_types() const;
+
+    bool last() const;
+
+    std::size_t column_count() const;
+
+    std::size_t row_count() const;
+
+    const std::vector<sql_row>& rows() const;
+
+private:
+    friend class sql_result;
+
+    std::vector<sql_column_type> column_types_;
+    std::vector<column> columns_;
+    std::vector<sql_row> rows_;
+    bool last_;
+    const sql_row_metadata* row_metadata_ = nullptr;
+    serialization::pimpl::SerializationService* serialization_service_;
+
+    template<typename T>
+    boost::optional<T> get_column_value(std::size_t column_index,
+                                        std::size_t row_index) const
+    {
+        assert(column_index < column_count());
+        assert(row_index < row_count());
+
+        auto& any_value = columns_[column_index][row_index];
+        if (any_value.empty()) {
+            return boost::none;
+        }
+
+        if (column_types_[column_index] != sql_column_type::object) {
+            return boost::any_cast<T>(any_value);
+        }
+
+        // this is the object type, hence the value is `data`
+        // and we need to de-serialize it
+        return serialization_service_->to_object<T>(
+          boost::any_cast<serialization::pimpl::data>(any_value));
+    }
+
+    /**
+     * set the row metadata on the page
+     * @param row_meta the row metadata pointer
+     */
+    void row_metadata(const sql_row_metadata* row_meta);
```

Comment:
That is a private field. I needed it because, on every fetch operation a page is retrieved but it does not have receive sql_row_metada on the response but we re-use the row metadata received from the initial sql execute response. And, we also need the metadata as part of the `sql_row` API, that is why we need it to be set on the page and from where we populate the metadata pointer in the sql_row. We have `page_iterator` which is different from the iterator of Java and I believe that is why it is different.

Reply:
ok makes sense. I thought this was public.